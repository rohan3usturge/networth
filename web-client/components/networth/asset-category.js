import { Asset } from "./asset";

const AssetCategory = ({ category, onChange }) => {
  const handleChange = (changedItem) => {
    const changed = { ...category };
    const newItems = category.items.map((item) => {
      if (item.lineItem.id === changedItem.lineItem.id) {
        return changedItem;
      }
      return item;
    });
    changed.items = newItems;
    if (onChange) {
      onChange(changed);
    }
  };

  return (
    <table className="table table-borderless">
      <thead>
        <tr className="d-flex">
          <th scope="col" className="col-4">
            {category.meta.name}
          </th>
          <th className="col-8"></th>
        </tr>
      </thead>
      <tbody>
        {category.items.map((item) => {
          return (
            <Asset
              key={item.lineItem.id}
              asset={item}
              onChange={handleChange}
            />
          );
        })}
      </tbody>
    </table>
  );
};

export { AssetCategory };
