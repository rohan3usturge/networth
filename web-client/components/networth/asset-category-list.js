import { NUMBER_UTILS } from "../utils";
import { AssetCategory } from "./asset-category";
import { useRouter } from "next/router";
import { withTranslation } from "react-i18next";

const Wrapper = ({ categories, onChange, disabled, total, currency, t }) => {
  const locale = useRouter().locale;
  const handleChange = (changedCategory) => {
    const changed = categories.map((category) => {
      if (category.meta.id == changedCategory.meta.id) {
        return changedCategory;
      }
      return category;
    });
    if (onChange) {
      onChange(changed);
    }
  };

  return (
    <>
      <div className="card mb-3">
        <div className="card-header text-white bg-success">
          <h4 className="card-title">
            {NUMBER_UTILS.convertNumToMoney(total, currency, locale)}
          </h4>
          <span className="card-text">
            <i className="fas fa-piggy-bank mr-2"></i>
            {t("assets")}
          </span>
        </div>
        <div className="card-body">
          {categories.map((category) => {
            return (
              <AssetCategory
                currency={currency}
                key={category.meta.id}
                category={category}
                onChange={handleChange}
                disabled={disabled}
              />
            );
          })}
        </div>
      </div>
    </>
  );
};

const AssetCategoryList = withTranslation()(Wrapper);

export { AssetCategoryList, Wrapper };
