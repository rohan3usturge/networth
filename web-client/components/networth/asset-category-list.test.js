import { shallow } from "enzyme";
import { Wrapper } from "./asset-category-list";
import { AssetCategory } from "./asset-category";

jest.mock("next/router", () => {
  return {
    useRouter: () => ({ locale: "en-ca" }),
  };
});

describe("AssetCategoryList", () => {
  it("should render properly and match snapshot", () => {
    const translations = {
      currency: "Currency",
    };

    const wrapper = shallow(
      <Wrapper
        categories={[]}
        currency="CAD"
        disabled={true}
        onChange={() => {}}
        total="2323.0232"
        key="2323"
        t={(key) => translations["currency"]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should call onChage correctly", () => {
    const translations = {
      currency: "Currency",
    };

    const categories = [
      {
        meta: {
          id: "1",
          name: "One",
        },
      },
      {
        meta: {
          id: "2",
          name: "Two",
        },
      },
    ];

    const onChange = jest.fn((changed) => {
      expect(changed).toHaveLength(2);
      expect(changed[0].meta.name).toBe("Three");
    });

    const wrapper = shallow(
      <Wrapper
        categories={categories}
        currency="CAD"
        disabled={true}
        onChange={onChange}
        total="2323.0232"
        key="2323"
        t={(key) => translations["currency"]}
      />
    );

    const props = wrapper.find(AssetCategory).first().props();
    const changeFunctionCall = props.onChange;
    changeFunctionCall({
      meta: {
        id: "1",
        name: "Three",
      },
    });
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it("should not call onChage if not pased", () => {
    const translations = {
      currency: "Currency",
    };

    const categories = [
      {
        meta: {
          id: "1",
          name: "One",
        },
      },
      {
        meta: {
          id: "2",
          name: "Two",
        },
      },
    ];

    const wrapper = shallow(
      <Wrapper
        categories={categories}
        currency="CAD"
        disabled={true}
        total="2323.0232"
        key="2323"
        t={(key) => translations["currency"]}
      />
    );

    const props = wrapper.find(AssetCategory).first().props();
    const changeFunctionCall = props.onChange;
    changeFunctionCall({
      meta: {
        id: "1",
        name: "Three",
      },
    });
  });
});
