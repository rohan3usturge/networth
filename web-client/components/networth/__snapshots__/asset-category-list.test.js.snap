// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetCategoryList should render properly and match snapshot 1`] = `
<Fragment>
  <div
    className="card mb-3"
  >
    <div
      className="card-header text-white bg-success"
    >
      <h4
        className="card-title"
      >
        CA$2,323.02
      </h4>
      <span
        className="card-text"
      >
        <i
          className="fas fa-piggy-bank mr-2"
        />
        Currency
      </span>
    </div>
    <div
      className="card-body"
    />
  </div>
</Fragment>
`;
